Bottom: f150652ca82a5866f82c590caaa5312f4a3a62bf
Top:    518dce469fbaa9c1ae345cdf4610d940f0283c41
Author: Ian Kent <ikent@redhat.com>
Date:   2013-04-04 12:56:54 +0800

autofs-5.0.7 - syncronize handle_mounts() shutdown


---

diff --git a/daemon/automount.c b/daemon/automount.c
index 4c651cf..476ff14 100644
--- a/daemon/automount.c
+++ b/daemon/automount.c
@@ -85,8 +85,8 @@ pthread_attr_t th_attr_detached;
 struct master_readmap_cond mrc = {
 	PTHREAD_MUTEX_INITIALIZER, PTHREAD_COND_INITIALIZER, 0, NULL, 0, 0, 0, 0};
 
-struct startup_cond suc = {
-	PTHREAD_MUTEX_INITIALIZER, PTHREAD_COND_INITIALIZER, 0, 0};
+struct shutdown_cond fc = {
+	PTHREAD_MUTEX_INITIALIZER, PTHREAD_COND_INITIALIZER, 0, 0, 0};
 
 pthread_key_t key_thread_stdenv_vars;
 
@@ -1338,6 +1338,12 @@ static void *statemachine(void *arg)
 		sigwait(&signalset, &sig);
 
 		switch (sig) {
+		case SIGCONT:
+			master_mutex_lock();
+			master_done(master_list);
+			master_mutex_unlock();
+			break;
+
 		case SIGTERM:
 		case SIGINT:
 		case SIGUSR2:
@@ -1348,7 +1354,8 @@ static void *statemachine(void *arg)
 					return NULL;
 				}
 			} else {
-				if (master_done(master_list)) {
+				master_done(master_list);
+				if (list_empty(&master_list->mounts)) {
 					master_mutex_unlock();
 					return NULL;
 				}
@@ -1443,6 +1450,47 @@ void handle_mounts_startup_cond_destroy(void *arg)
 	return;
 }
 
+void finish_mutex_lock(void)
+{
+	int status = pthread_mutex_lock(&fc.mutex);
+	if (status)
+		logerr("failed to lock shutdown condition mutex!");
+		fatal(status);
+	}
+}
+
+void finish_mutex_unlock(void)
+{
+	int status = pthread_mutex_unlock(&fc.mutex);
+	if (status)
+		logerr("failed to unlock shutdown condition mutex!");
+		fatal(status);
+	}
+}
+
+void finish_cond_wait(void)
+{
+	int status = pthread_cond_wait(&fc.cond, &fc.mutex);
+	if (status)
+		fatal(status);
+}
+
+void handle_mounts_done(void)
+{
+	int status, cancel_state;
+
+	pthread_setcancelstate(PTHREAD_CANCEL_DISABLE, &cancel_state);
+
+	finish_mutex_lock();
+	sdc.busy++;
+	/* Poke signal handler */
+	pthread_kill(state_mach_thid, SIGCONT);
+	finish_cond_wait();
+	finish_mutex_unlock();
+
+	pthread_setcancelstate(cancel_state, NULL);
+}
+
 static void handle_mounts_cleanup(void *arg)
 {
 	struct autofs_point *ap;
@@ -1493,15 +1541,15 @@ static void handle_mounts_cleanup(void *arg)
 
 	info(logopt, "shut down path %s", path);
 
+	master_mutex_unlock();
+
 	/*
 	 * If we are not a submount send a signal to the signal handler
 	 * so it can join with any completed handle_mounts() threads and
 	 * perform final cleanup.
 	 */
 	if (!submount)
-		pthread_kill(state_mach_thid, SIGTERM);
-
-	master_mutex_unlock();
+		handle_mounts_done();
 
 	return;
 }
diff --git a/include/automount.h b/include/automount.h
index e72fa0d..565adad 100644
--- a/include/automount.h
+++ b/include/automount.h
@@ -353,6 +353,17 @@ struct startup_cond {
 int handle_mounts_startup_cond_init(struct startup_cond *suc);
 void handle_mounts_startup_cond_destroy(void *arg);
 
+struct finish_cond {
+	pthread_mutex_t mutex;
+	pthread_cond_t  cond;
+	unsigned int busy;
+};
+
+extern struct finish_cond fc;
+void finish_mutex_lock(void);
+void finish_mutex_unlock(void);
+void finish_cond_wait(void);
+
 struct master_readmap_cond {
 	pthread_mutex_t mutex;
 	pthread_cond_t  cond;
diff --git a/include/master.h b/include/master.h
index 824707c..8913740 100644
--- a/include/master.h
+++ b/include/master.h
@@ -115,7 +115,7 @@ int master_mount_mounts(struct master *, time_t, int);
 int master_show_mounts(struct master *);
 extern inline unsigned int master_get_logopt(void);
 int master_list_empty(struct master *);
-int master_done(struct master *);
+void master_done(struct master *);
 int master_kill(struct master *);
 
 #endif
diff --git a/lib/master.c b/lib/master.c
index a0e62f2..491fda0 100644
--- a/lib/master.c
+++ b/lib/master.c
@@ -1394,26 +1394,33 @@ int master_list_empty(struct master *master)
 	return res;
 }
 
-int master_done(struct master *master)
+void master_done(struct master *master)
 {
 	struct list_head *head, *p;
 	struct master_mapent *entry;
-	int res = 0;
+	int status;
 
-	head = &master->completed;
-	p = head->next;
-	while (p != head) {
-		entry = list_entry(p, struct master_mapent, join);
-		p = p->next;
-		list_del(&entry->join);
-		pthread_join(entry->thid, NULL);
-		master_free_mapent_sources(entry, 1);
-		master_free_mapent(entry);
+	finish_mutex_lock();
+
+	while (sdc.busy) {
+		head = &master->completed;
+		p = head->next;
+		while (p != head) {
+			entry = list_entry(p, struct master_mapent, join);
+			p = p->next;
+			list_del(&entry->join);
+			pthread_join(entry->thid, NULL);
+			master_free_mapent_sources(entry, 1);
+			master_free_mapent(entry);
+			sdc.busy--;
+		}
 	}
-	if (list_empty(&master->mounts))
-		res = 1;
 
-	return res;
+	status = pthread_cond_broadcast(&fc);
+	if (status)
+		fatal(status);
+
+	finish_mutex_unlock();
 }
 
 inline unsigned int master_get_logopt(void)
